client/rpc-servers/src/lib.rs

```diff
@@ -20,213 +20,193 @@
+) -> Result<HttpServerHandle, Box<dyn StdError + Send + Sync>> {
+	let max_payload_in = payload_size_or_default(max_payload_in_mb);
+	let max_payload_out = payload_size_or_default(max_payload_out_mb);
+
+	let mut acl = AccessControlBuilder::new();
+
+	if let Some(cors) = cors {
+		// Whitelist listening address.
+		// NOTE: set_allowed_hosts will whitelist both ports but only one will used.
+		acl = acl.set_allowed_hosts(format_allowed_hosts(&addrs[..]))?;
+		acl = acl.set_allowed_origins(cors)?;
+	};
+
+	let builder = HttpServerBuilder::new()
+		.max_request_body_size(max_payload_in as u32)
+		.max_response_body_size(max_payload_out as u32)
+		.set_access_control(acl.build())
+		.custom_tokio_runtime(rt);
+
+	let rpc_api = build_rpc_api(rpc_api);
+	let (handle, addr) = if let Some(metrics) = metrics {
+		let middleware = RpcMiddleware::new(metrics, "http".into());
+		let builder = builder.set_middleware(middleware);
+		let server = builder.build(&addrs[..]).await?;
+		let addr = server.local_addr();
+		(server.start(rpc_api)?, addr)
+	} else {
+		let server = builder.build(&addrs[..]).await?;
+		let addr = server.local_addr();
+		(server.start(rpc_api)?, addr)
+	};
+
+	log::info!(
+		"Running JSON-RPC HTTP server: addr={}, allowed origins={:?}",
+		addr.map_or_else(|_| "unknown".to_string(), |a| a.to_string()),
+		cors
+	);
+
+	Ok(handle)
 }
 
 /// Start WS server listening on given address.
-pub fn start_ws<
-	M: pubsub::PubSubMetadata + From<futures::channel::mpsc::UnboundedSender<String>>,
->(
-	addr: &std::net::SocketAddr,
-	max_connections: Option<usize>,
+pub async fn start_ws<M: Send + Sync + 'static>(
+	addrs: [SocketAddr; 2],
 	cors: Option<&Vec<String>>,
-	io: RpcHandler<M>,
-	maybe_max_payload_mb: Option<usize>,
-	maybe_max_out_buffer_capacity_mb: Option<usize>,
-	server_metrics: ServerMetrics,
-	tokio_handle: tokio::runtime::Handle,
-) -> io::Result<ws::Server> {
-	let max_payload = maybe_max_payload_mb
-		.map(|mb| mb.saturating_mul(MEGABYTE))
-		.unwrap_or(RPC_MAX_PAYLOAD_DEFAULT);
-	let max_out_buffer_capacity = maybe_max_out_buffer_capacity_mb
-		.map(|mb| mb.saturating_mul(MEGABYTE))
-		.unwrap_or(WS_MAX_BUFFER_CAPACITY_DEFAULT);
-
-	if max_payload > max_out_buffer_capacity {
-		log::warn!(
-			"maximum payload ({}) is more than maximum output buffer ({}) size in ws server, the payload will actually be limited by the buffer size",
-			max_payload,
-			max_out_buffer_capacity,
-		)
+	ws_config: WsConfig,
+	metrics: Option<RpcMetrics>,
+	rpc_api: RpcModule<M>,
+	rt: tokio::runtime::Handle,
+	id_provider: Option<Box<dyn IdProvider>>,
+) -> Result<WsServerHandle, Box<dyn StdError + Send + Sync>> {
+	let (max_payload_in, max_payload_out, max_connections, max_subs_per_conn) =
+		ws_config.deconstruct();
+
+	let mut builder = WsServerBuilder::new()
+		.max_request_body_size(max_payload_in)
+		.max_response_body_size(max_payload_out)
+		.max_connections(max_connections)
+		.max_subscriptions_per_connection(max_subs_per_conn)
```

Comment:
heads up: breaking this is a way to limit max subscriptions per connection by default it is 1024 but also configurable via the CLI.

Reply:
How do you mean breaking? Do you mean that before there was no limit (other than the `jsonrpc` buffers running out and dropping the connection)? That is fine I think. 1024 seems like a high enough number.

Reply:
good point, I meant that we didn't have this functionality before in jsonrpc. Just poorly phrased